lib/queuesender/sender.go

```diff
@@ -0,0 +1,190 @@
+package queuesender
+
+import (
+	"fmt"
+	"github.com/Symantec/tricorder/go/tricorder"
+	"github.com/Symantec/tricorder/go/tricorder/units"
+	"log"
+)
+
+func _new(endpoint string, size int, name string, logger *log.Logger) (
+	*Sender, error) {
+	if size < 1 {
+		panic("Size must be at least 1")
+	}
+	conn, err := newConn(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	result := &Sender{
+		queue:        newQueue(size),
+		newConnCh:    make(chan *connType),
+		newConnReqCh: make(chan bool),
+	}
+	if logger != nil {
+		result.logger = &loggerType{name: name, logger: logger}
+	}
+	go result.sendLoop(conn)
+	go result.receiveLoop(conn)
+	return result, nil
+}
+
+// This loop sends requests in the queue. To coordinate connection refreshes
+// between the send loop and receive loop goroutines, only this send loop
+// closes stale connections, and only the receiving loop refreshes connections.
+// Whenever this send loop closes a stale connection, the receiving loop has
+// moved on to using a new connection.
+func (s *Sender) sendLoop(conn *connType) {
+	connId := 0
+	for {
+		// See if the receiving loop has a new connection for us.
+		// If not, just keep using the same connection
+		select {
+		case newConn := <-s.newConnCh:
+			// We know its safe to close the old connection because
+			// the receiving loop is already using the connection it
+			// just gave us.
+			conn.Close()
+			conn = newConn
+			// For a given connection, both the send and receiving loop
+			// use the same value for connection ID.
+			connId++
+		default:
+		}
+		req, ok := s.queue.NextNotSent(connId)
+		if !ok {
+			// If we get here, the receiver loop has already created
+			// a new connection, and the queue is already tracking it.
+			// Go to the beginning to get this new connection.
+			continue
+		}
+		err := conn.Send(req)
```

Comment:
If the reader loop refreshes the connection and the sender is blocked here, no progress will be made.

When the reader loop refreshes the connection, the sender loop needs to be interrupted here. Maybe set some sort of timeout on the connection?